File: _system_scripts/organize_selection.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
📁 エピソード整理スクリプト (Refactored)
"""

import sys
from pathlib import Path

# プロジェクトルートをパスに追加
sys.path.append(str(Path(__file__).resolve().parent.parent))

import re
import shutil
import glob
from lib import config, utils

# 初期化
logger = utils.initialize_script("organize_selection")

# 章番号とフォルダ名のマッピング
CHAPTER_MAP = {
    1: "01_現状把握と脱労働",
    2: "02_投資の魔法と基礎",
    3: "03_実践と準備",
    4: "04_継続の技術",
    5: "05_投資の果実と自由"
}

# 同名回避のためのキーワード
TITLE_KEYWORDS = {
    1: "配当貴族",
}

def parse_selection_list():
    content = utils.FileIO.read_text(config.SELECTION_LIST_FILE)
    if content is None: return []
    
    selections = []
    current_chapter = 0
    
    for line in content.splitlines():
        # Chapter検出
        m_chap = re.match(r"^## 第(\d+)章", line)
        if m_chap:
            current_chapter = int(m_chap.group(1))
            continue
        
        # テーブル行検出
        if not line.strip().startswith('|'):
            continue
        
        parts = [p.strip() for p in line.strip().split('|')]
        if len(parts) < 4: continue
        
        no_str = parts[1]
        title_str = parts[2]
        
        if not no_str.isdigit():
            continue
            
        no = int(no_str)
        title = title_str.replace('*', '').strip()
        
        selections.append({
            'chapter': current_chapter,
            'no': no,
            'title': title
        })
    return selections

def main():
    selections = parse_selection_list()
    logger.info("Loaded %d episodes from selection list.", len(selections))
    
    # ターゲットディレクトリの作成
    for folder in CHAPTER_MAP.values():
        path = config.MANGA_DIR / folder
        path.mkdir(parents=True, exist_ok=True)

    files_moved = 0
    for item in selections:
        folder_name = CHAPTER_MAP.get(item['chapter'])
        if not folder_name:
            logger.warning("Unknown chapter %d for No.%d", item['chapter'], item['no'])
            continue
            
        target_dir = config.MANGA_DIR / folder_name
        pattern = f"No{item['no']:02d}_*プロンプト.md"
        candidates = glob.glob(str(config.MANGA_DIR / "**" / pattern), recursive=True)
        
        # フィルタリング
        valid_candidates = []
        keyword = TITLE_KEYWORDS.get(item['no'])
        
        for p in candidates:
            if keyword and keyword not in Path(p).name:
                continue
            valid_candidates.append(p)
            
        if not valid_candidates:
            logger.debug("No.%d (%s) - File NOT FOUND.", item['no'], item['title'])
            continue
            
        src_path = valid_candidates[0]
        if len(valid_candidates) > 1:
            for vc in valid_candidates:
                if item['title'] in Path(vc).name:
                    src_path = vc
                    break
        
        src_path = Path(src_path)
        dest_path = target_dir / src_path.name
        
        if src_path.resolve() == dest_path.resolve():
            logger.debug("OK (Already in place): %s", src_path.name)
            continue
            
        try:
            shutil.move(str(src_path), str(dest_path))
            logger.info("MOVED: %s -> %s", src_path.name, folder_name)
            files_moved += 1
        except Exception as e:
            logger.error("ERROR moving %s: %s", src_path.name, e)

    logger.info("Operation Complete. Moved %d files.", files_moved)

if __name__ == "__main__":
    main()

File: _system_scripts/old_tools/organize_selection.js
const fs = require('fs');
const path = require('path');

const ROOT_DIR = "c:\\Users\\hirak\\Desktop\\2nd-Brain\\18_レミ投資漫画";

const CHAPTER_MAP = {
    1: "01_現状把握と脱労働",
    2: "02_投資の魔法と基礎",
    3: "03_実践と準備",
    4: "04_継続の技術",
    5: "05_投資の果実と自由"
};

const TITLE_KEYWORDS = {
    1: "配当貴族"
};

function parseSelectionList() {
    const selectionFile = path.join(ROOT_DIR, "05_設定", "初心者向け連載30選.md");
    const selections = [];
    let currentChapter = 0;

    try {
        const content = fs.readFileSync(selectionFile, 'utf8');
        const lines = content.split(/\r?\n/);

        for (const line of lines) {
            const chapMatch = line.match(/^## 第(\d+)章/);
            if (chapMatch) {
                currentChapter = parseInt(chapMatch[1], 10);
                continue;
            }

            if (!line.trim().startsWith('|')) continue;

            const parts = line.split('|').map(p => p.trim());
            if (parts.length < 4) continue;

            const noStr = parts[1];
            const titleStr = parts[2];

            if (!/^\d+$/.test(noStr)) continue;

            const no = parseInt(noStr, 10);
            const title = titleStr.replace(/\*/g, '').trim();

            selections.push({
                chapter: currentChapter,
                no: no,
                title: title
            });
        }
    } catch (e) {
        console.error(`Error reading selection list: ${e.message}`);
    }
    return selections;
}

function findFile(no, titleKeyword) {
    let found = [];

    function walk(dir) {
        const files = fs.readdirSync(dir);
        for (const f of files) {
            const fullPath = path.join(dir, f);
            const stat = fs.statSync(fullPath);
            if (stat.isDirectory()) {
                walk(fullPath);
            } else {
                // Check pattern: NoXX_*プロンプト.md or similar
                // We need to be careful. Let's look for "No" + numbers + something + "プロンプト.md"
                // The python script used: f"No{item['no']:02d}_*プロンプト.md"
                // JS regex: /No(0?1)_.*プロンプト\.md/

                // Construct regex for this specific No
                // valid formats: No1_..., No01_..., No.1_...
                // We'll stick to simple check: file starts with No+number

                // Matches "No" followed by the number (with optional 0 padding) and a non-digit
                // e.g. for No=1: No01_..., No1_...
                const nameHelper = `No${no.toString().padStart(2, '0')}_`;
                const nameHelper2 = `No${no}_`;

                if ((f.startsWith(nameHelper) || f.startsWith(nameHelper2)) && f.endsWith('プロンプト.md')) {
                    // If keyword restriction exists
                    if (titleKeyword) {
                        if (!f.includes(titleKeyword)) return;
                    }
                    found.push(fullPath);
                }
            }
        }
    }

    try {
        walk(ROOT_DIR);
    } catch (e) {
        console.error(`Error walking directory: ${e.message}`);
    }
    return found;
}

function main() {
    console.log("Starting organization...");
    const selections = parseSelectionList();
    console.log(`Loaded ${selections.length} episodes.`);

    // Ensure target dirs exist
    for (const key in CHAPTER_MAP) {
        const targetDir = path.join(ROOT_DIR, CHAPTER_MAP[key]);
        if (!fs.existsSync(targetDir)) {
            fs.mkdirSync(targetDir, { recursive: true });
            console.log(`Created dir: ${CHAPTER_MAP[key]}`);
        }
    }

    let movedCount = 0;

    for (const item of selections) {
        const targetFolderName = CHAPTER_MAP[item.chapter];
        if (!targetFolderName) {
            console.warn(`Unknown chapter ${item.chapter} for No.${item.no}`);
            continue;
        }

        const targetDir = path.join(ROOT_DIR, targetFolderName);
        const keyword = TITLE_KEYWORDS[item.no];

        let candidates = findFile(item.no, keyword);

        if (candidates.length === 0) {
            console.log(`File NOT FOUND for No.${item.no} (${item.title})`);
            continue;
        }

        let srcPath = candidates[0];
        // If multiple, try to match title
        if (candidates.length > 1) {
            const exactMatch = candidates.find(c => path.basename(c).includes(item.title));
            if (exactMatch) srcPath = exactMatch;
        }

        const fileName = path.basename(srcPath);
        const destPath = path.join(targetDir, fileName);

        if (path.resolve(srcPath) === path.resolve(destPath)) {
            console.log(`Already in place: ${fileName}`);
            continue;
        }

        try {
            fs.renameSync(srcPath, destPath);
            console.log(`MOVED: ${fileName} -> ${targetFolderName}`);
            movedCount++;
        } catch (e) {
            console.error(`ERROR moving ${fileName}: ${e.message}`);
        }
    }

    console.log(`\nDone. Moved ${movedCount} files.`);
}

main();
